File: go/search/leetcode/word_ladder.go
package leetcode

//单词接龙

//广度优先
func ladderLength(beginWord string, endWord string, wordList []string) int {
	wordSet, charSet := make(map[string]bool), make(map[rune]bool)
	for _, word := range wordList {
		wordSet[word] = true
		for _, c := range word {
			charSet[c] = true
		}
	}
	if !wordSet[endWord] {
		return 0
	}

	queue, res := []string{beginWord}, 0
	for len(queue) > 0 {
		n := len(queue)
		for i := 0; i < n; i++ {
			cur := queue[0]
			queue = queue[1:]
			if cur == endWord {
				return res + 1
			}
			for j := 0; j < len(cur); j++ {
				for c := range charSet {
					modStr := cur[:j] + string(c) + cur[j+1:]
					if wordSet[modStr] {
						queue = append(queue, modStr)
						delete(wordSet, modStr)
					}
				}
			}
			res++
		}
	}
	return 0
}

File: python/search/leetcode/word_ladder.py
from typing import List


# 单词接龙

class Solution:

    # 广度优先搜索的思路
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordSet = set(wordList)
        if endWord not in wordSet:
            return 0
        
        queue, res = [beginWord], 0
        while len(queue) > 0:
            n = len(queue)
            for i in range(n):
                cur = queue[0]
                queue = queue[1:]
                if cur == endWord:
                    return res + 1
                
                for i in range(len(cur)):
                    for c in range(97, 97 + 26):
                        modStr = cur[:i] + chr(c) + cur[i + 1:]
                        if modStr in wordSet:
                            queue.append(modStr)
                            wordSet.remove(modStr)
            res += 1
        return 0
